test/e2e/storage/non_csi_volumes.go

```diff
@@ -0,0 +1,1170 @@
+	"subPath":                          initSubPathTestCase,
+}
+
+var _ = utils.SIGDescribe("Non-CSI", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
+		config framework.VolumeTestConfig
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+		nodes := framework.GetReadySchedulableNodesOrDie(cs)
+		node = nodes.Items[rand.Intn(len(nodes.Items))]
+		config = framework.VolumeTestConfig{
+			Namespace:      ns.Name,
+			Prefix:         "volume",
+			ClientNodeName: node.Name,
+			ServerNodeName: node.Name,
+		}
+	})
+
+	for driverName, initDriver := range testDrivers {
+		curDriverName := driverName
+		curInitDriver := initDriver
+
+		for testCaseName, testCaseInit := range testCases {
+			curTestCaseName := testCaseName
+			curTestCaseInit := testCaseInit
+			curTestCase := curTestCaseInit()
+			curTestPatterns := curTestCase.getTestPatterns()
+			curTestFunc := curTestCase.getTestFunc()
+
+			Context(fmt.Sprintf("Driver: %s, Test: %s", curDriverName, curTestCaseName), func() {
+				var (
+					driver   testDriver
+					testCase testCase
+				)
+				BeforeEach(func() {
+					testCase = curTestCase
+					driver = curInitDriver(f, config, node)
+					driver.createDriver()
+				})
+
+				AfterEach(func() {
+					driver.cleanupDriver()
+				})
+
+				for testPatternName, testPattern := range curTestPatterns {
+					curTestPatternName := testPatternName
+					curTestPattern := testPattern
+
+					Context(fmt.Sprintf("TestPattern: %s", curTestPatternName), func() {
+						var (
+							testResource testResource
+							testInput    testInput
+						)
+
+						BeforeEach(func() {
+							testResource = testCase.getTestResource()
+							testResource.init(driver, curTestPattern)
+							testInput = testCase.createTestInput(curTestPattern, testResource)
+						})
+
+						AfterEach(func() {
+							testResource.cleanup()
+						})
+
+						curTestFunc(&testInput)
+					})
+				}
+			})
+		}
+	}
+})
+
+type testDriver interface {
+	createDriver()
+	cleanupDriver()
+	getDriverInfo() driverInfo
+	getVolumeSource(readOnly bool) *v1.VolumeSource
+	getPersistentVolumeSource(readOnly bool) *v1.PersistentVolumeSource
+}
+
+type driverInfo struct {
+	name                      string
+	fsGroup                   *int64
+	podSec                    v1.PodSecurityContext
+	privilegedSecurityContext bool
+	maxFileSize               int64
+
+	// supportablity for driver
+	isBlockSupported bool
+
+	// volumes test specific parameters
+	expectedContent string
+	needInjectHTML  bool
+
+	// volumeIO test specific parameter
+	testFile string
+
+	f      *framework.Framework
+	config framework.VolumeTestConfig
+	node   v1.Node
+}
+
+type testCase interface {
+	getTestPatterns() map[string]testPattern
+	getTestResource() testResource
+	createTestInput(testPattern, testResource) testInput
+	getTestFunc() func(*testInput)
+}
+
+type testPattern struct {
+	isPVtest    bool
```

Comment:
these parameters are not used by all tests.  Can each test suite implement their own type that implements an Interface?  Then you can store test-specific parameters in the specific struct.